qiskit/circuit/classical/expr/constructors.py

```diff
@@ -181,11 +188,16 @@ def logic_not(operand: typing.Any, /) -> Expr:
             >>> expr.logic_not(ClassicalRegister(3, "c"))
             Unary(\
 Unary.Op.LOGIC_NOT, \
-Cast(Var(ClassicalRegister(3, 'c'), Uint(3)), Bool(), implicit=True), \
+Cast(Var(ClassicalRegister(3, 'c'), Uint(3)), \
+Bool(), implicit=True), \
 Bool())
     """
-    operand = _coerce_lossless(lift(operand), types.Bool())
-    return Unary(Unary.Op.LOGIC_NOT, operand, operand.type)
+    operand = lift(operand)
+    try:
+        operand = _coerce_lossless(operand, types.Bool())
+        return Unary(Unary.Op.LOGIC_NOT, operand, operand.type)
+    except TypeError as ex:
+        raise TypeError(f"cannot apply '{Unary.Op.BIT_NOT}' to type '{operand.type}'") from ex
```

Comment:
only the `_coerce_lossless` can fail, so if you want to modify its exception
```python
try:
    operand = _coerce_lossless(operand, types.Bool())
except TypeError:
    raise TypeError(...) from None
return Unary(..., operand, operand.type)
```
might be a bit clearer about the scope of the `try`.

As minor nitpicks too:
- the chained exception is an internal implementation detail, so it's potentially cleaner for the user to suppress chaining with `raise TypeError(...) from None`, otherwise it kind of looks like an internal mistake (imo)
- `Unary.Op.LOGIC_NOT` is probably too low-level detail for a user of this function (also you have a typo in the operand) - it's maybe clearer to say "cannot apply 'logic_not' to type ..." so the link is clearer

Reply:
I'll make these changes, but I think we ought to use `Unary.Op.LOGIC_NOT` in the message since that's what the other constructors do.

Reply:
Ok sounds fine if there's a consistency argument.  Now you mention it, I think I used the `Binary.Op` thing mostly just out of ease.